File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.py
class Solution:
    def getDp(self, dp, i, j):
        if i > j:
            return 0
        return dp[i][j]

    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        dp = [[0] * n for _ in range(n)]

        for a in range(0, n):
            j = a
            i = 0
            while( j < n ):
                if s[i] == s[j]:
                    score = 1 if i == j else 2
                    if dp[i][j] < (score + self.getDp(dp, i+1, j-1)):
                        dp[i][j] = score + self.getDp(dp, i+1, j-1)

                if dp[i][j] < self.getDp(dp, i+1, j):
                    dp[i][j] = max(dp[i][j], self.getDp(dp, i+1, j))
                if dp[i][j] < self.getDp(dp, i, j - 1):
                    dp[i][j] = max(dp[i][j], self.getDp(dp, i, j-1))
                i += 1
                j += 1
        return dp[0][n-1]

File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.cpp
class Solution {
public:
    int longestPalindromeSubseq(string s) {
        int n = s.size();

        vector<vector<int>> dp (n+1, vector<int>(n+1, 0));

        for(int i = 1; i < n+1; i++){
            dp[i][i] = 1;
        }

        for(int i = n-1; i >= 1; i--){
            for(int j = i+1; j < n+1; j++) {
                dp[i][j] = max(dp[i+1][j], dp[i][j-1]);
                if(s[i-1] == s[j-1]){
                    dp[i][j] = max(dp[i][j], 2 + dp[i+1][j-1]);
                }
            }
        }
        return dp[1][n];
    }
};